pkg/traceql/engine_metrics_average.go

```diff
@@ -0,0 +1,540 @@
+		return a.agg.Series()
+	}
+
+	// In the frontend-version the results come from
+	// the job-level aggregator
+	ss := a.seriesAgg.Results()
+	if a.mode == AggregateModeFinal {
+		for i := range ss {
+			if strings.Contains(i, "_type") {
+				delete(ss, i)
+			}
+		}
+	}
+	return ss
+}
+
+func (a *AverageOverTimeAggregator) extractConditions(request *FetchSpansRequest) {
+	// For metrics aggregators based on a span attribute we have to include it
+	includeAttribute := a.attr != (Attribute{}) && !request.HasAttribute(a.attr)
+	if includeAttribute {
+		request.SecondPassConditions = append(request.SecondPassConditions, Condition{
+			Attribute: a.attr,
+		})
+	}
+
+	for _, b := range a.by {
+		if !request.HasAttribute(b) {
+			request.SecondPassConditions = append(request.SecondPassConditions, Condition{
+				Attribute: b,
+			})
+		}
+	}
+}
+
+func (a *AverageOverTimeAggregator) validate() error {
+	if len(a.by) >= maxGroupBys {
+		return newUnsupportedError(fmt.Sprintf("metrics group by %v values", len(a.by)))
+	}
+	return nil
+}
+
+func (a *AverageOverTimeAggregator) spanStartTimeMs(s Span) uint64 {
+	return s.StartTimeUnixNanos() / uint64(time.Millisecond)
+}
+
+func (a *AverageOverTimeAggregator) String() string {
+	s := strings.Builder{}
+
+	s.WriteString(metricsAggregateAvgOverTime.String())
+	s.WriteString("(")
+	if a.attr != (Attribute{}) {
+		s.WriteString(a.attr.String())
+	}
+	s.WriteString(")")
+
+	if len(a.by) > 0 {
+		s.WriteString("by(")
+		for i, b := range a.by {
+			s.WriteString(b.String())
+			if i < len(a.by)-1 {
+				s.WriteString(",")
+			}
+		}
+		s.WriteString(")")
+	}
+	return s.String()
+}
+
+type SpanSetsAverageOverTimeAggregator struct {
+	ss               SeriesSet
+	countProm        map[string]string
+	len              int
+	start, end, step uint64
+	exemplarBuckets  *bucketSet
+}
+
+var (
+	_   SeriesAggregator = (*SpanSetsAverageOverTimeAggregator)(nil)
+	nan                  = math.Float64frombits(normalNaN)
+)
+
+func (b *SpanSetsAverageOverTimeAggregator) Combine(in []*tempopb.TimeSeries) {
+	newCountersTS := make(map[string][]float64)
+
+	b.initSeriesAggregator(in, newCountersTS)
+	for _, ts := range in {
+		counterLabel, ok := b.countProm[ts.PromLabels]
+		if !ok {
+			// This is a counter label, we can skip it
+			continue
+		}
+		existing := b.ss[ts.PromLabels]
+		for _, sample := range ts.Samples {
+			pos := IntervalOfMs(sample.TimestampMs, b.start, b.end, b.step)
+			if pos < 0 || pos > len(b.ss[ts.PromLabels].Values) {
+				continue
+			}
+
+			currentAvg := b.ss[ts.PromLabels].Values[pos]
+			newAvg := sample.Value
+			currentCount := b.ss[counterLabel].Values[pos]
+			newCount := newCountersTS[ts.PromLabels][pos]
+
+			if math.IsNaN(currentAvg) && !math.IsNaN(newAvg) {
+				b.ss[ts.PromLabels].Values[pos] = newAvg
+				b.ss[counterLabel].Values[pos] = newCount
+			} else if !math.IsNaN(newAvg) {
+				// Weighted mean
+				avg := (currentAvg*currentCount + newAvg*newCount) / (currentCount + newCount)
```

Comment:
It calculates the weighted mean. There is another to reduce the intermediate operations (and a possible loss of precision) by adding all the values and counts and calculating the average on the Results()

Reply:
Now that I've reviewed further - could we reuse the kahan incremental average here too?  Across a few thousands of jobs, it seems like it could be worthwhile.

Reply:
Went ahead and did some basic testing of error rates, and kahan approach is definitely better.    Calculated the average of pi and 1e12 a million times each.  Kahan had zero error across either.  Simple sum and divide accumlated a small error on pi.

```
KahanSum: Input: 3.141592653589793 Avg: 3.141592653589793 Error: 0
KahanSum: Input: 1e+12 Avg: 1e+12 Error: 0

Simple: Input: 3.141592653589793 Avg: 3.1415926535849428 Error: 1.5438761380437427e-12
Simple: Input: 1e+12 Avg: 1e+12 Error: 0
```


Here is the code:
<details>

```go
func TestErrorKahanSum(t *testing.T) {
	fs := []float64{math.Pi, 1e12}

	for _, f := range fs {
		t.Run(strconv.FormatFloat(f, 'g', -1, 64), func(t *testing.T) {
			avg := 0.0
			count := 0.0
			c := 0.0

			for i := 0; i < 1_000_000; i++ {
				count++
				avg, c = averageInc(avg, f, count, c)
			}

			fmt.Println("KahanSum: Input:", f, "Avg:", avg, "Error:", 1.0-avg/f)
		})
	}
}

func TestErrorSimpleAvg(t *testing.T) {
	fs := []float64{math.Pi, 1e12}

	for _, f := range fs {
		t.Run(strconv.FormatFloat(f, 'g', -1, 64), func(t *testing.T) {
			sum := 0.0
			count := 0.0

			for i := 0; i < 1_000_000; i++ {
				count++
				sum += f
			}

			avg := sum / count

			fmt.Println("Simple: Input:", f, "Avg:", avg, "Error:", 1.0-avg/f)
		})
	}
}
```

</details>